### PR TITLE
Prototype MTDStageExample-based stage testing (#85)

### PR DESCRIPTION
**Pull request by mmcdermott**
## Summary

Migrates `meds-torch-data`'s preprocessing stage tests from a ~767-line bespoke harness to MEDS-Transforms 0.6.x's declarative `StageExample`-based machinery (tracked in #85). Prunes the old harness in the same PR.

## Approach

- **`MTDStageExample(StageExample)`** subclass (`src/meds_torchdata/preprocessing/_stage_example.py`, ~140 LoC). Overrides `is_example_dir`, `from_dir`, and `check_outputs` to validate declarative `out_data.yaml` specs that describe stage outputs from the cohort root. Dispatches per-path on suffix:
  - `.parquet` → `polars.testing.assert_frame_equal` (with `check_dtypes=False` since YAML can't express polars dtypes).
  - `.nrt` → `JointNestedRaggedTensorDict` tensor-dict diff via `np.array_equal(equal_nan=True)` per key.

- **Per-stage subpackage layout** mirroring the upstream convention (e.g. `MEDS_transforms/stages/reshard_to_split/`):
  ```
  preprocessing/
    tokenization/
      __init__.py            # from .tokenization import main
      tokenization.py        # @Stage.register'd function
      examples/
        default/             # two-subject baseline
        missing_static/      # subject with no static rows → null static lists
        multi_shard/         # four shards across three splits
    tensorization/           # same layout, default/ + multi_shard/
  ```
  With this layout, `Stage.register` auto-infers `examples_dir` — both decorators collapse to `@Stage.register(is_metadata=False, example_class=MTDStageExample)`. The `[project.entry-points."MEDS_transforms.stages"]` paths are unchanged because each subpackage's `__init__.py` re-exports `main`. MEDS-Transforms auto-docgen picks the stages up via the same machinery as upstream's built-ins.

- **Test entry point is one-line** (`tests/preprocessing/test_stages.py`):
  ```python
  def test_stage_example(stage_example):
      stage_example.test()
  ```
  Auto-parametrized across scenarios via `MEDS_transforms.pytest_plugin` (already entry-point registered via the `meds-transforms` package).

- **`pipeline_tester`** runs tokenization → tensorization end-to-end through `MEDS_transform-pipeline` via `test_pipeline_serial` (and `test_pipeline_parallel` under a joblib stage runner, gated on `@pytest.mark.parallelized`), validating each stage's output against its `default` scenario.

## Pruned

| File | Before | After |
|---|---|---|
| `tests/preprocessing/__init__.py` | 767 lines | 9 lines (just `PREPROCESS_SCRIPT`) |
| `tests/preprocessing/test_tokenization.py` | 457 lines | **deleted** |
| `tests/preprocessing/test_tensorization.py` | 29 lines | **deleted** |
| `tests/preprocessing/test_preprocess_parallel.py` | 64 lines | **deleted** |
| `tests/preprocessing/test_preprocess.py` | 113 lines | 97 lines (CLI-only) |

## Coverage notes

The three scenarios cover everything the old harness tested:

- **`default`** — two subjects, one shard, full static + dynamic data. Baseline.
- **`missing_static`** — a subject with no static rows (null static_code / static_numeric_value). Previously covered by `NORMALIZED_TRAIN_0_MISSING_STATIC` / `WANT_SCHEMAS_TRAIN_0_MISSING_STATIC`.
- **`multi_shard`** — four shards across three splits (`train/0`, `train/1`, `tuning/0`, `held_out/0`). Previously implicit in the old harness's `remap_inputs_for_transform` writing each shard to its own parquet.

The CLI wrapper's `stage_runner_fp=` hydra-override plumbing (previously covered by `test_preprocess_parallel.py`'s `MTD_preprocess stage_runner_fp=...` invocation) now has a dedicated `test_preprocess_stage_runner_fp_passthrough` in `test_preprocess.py`, since `test_pipeline_parallel` uses `pipeline_tester` which calls `MEDS_transform-pipeline` directly and bypasses our wrapper.

## Test plan

- [x] `uv run pytest tests/preprocessing/test_stages.py -v --no-cov -m "not parallelized"` — 6 passed (5 stage scenarios + 1 serial pipeline).
- [x] `uv run pytest tests/ --no-cov -m "not lightning"` (with `uv sync --extra parallelized`) — 33 passed, 2 skipped.
- [x] `uv build --wheel` — scenario YAMLs ship in the wheel via setuptools-scm.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/meds_torchdata/preprocessing/tokenization.py [hunk 1]
```diff
@@ -304,7 +306,11 @@ def extract_seq_of_subject_events(df: pl.LazyFrame) -> pl.LazyFrame:
     )
 
 
-@Stage.register(is_metadata=False)
+@Stage.register(
+    is_metadata=False,
+    example_class=MTDStageExample,
+    examples_dir=Path(__file__).parent / "tokenization_examples",
+)
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/preprocessing/tokenization.py]: Fixed in 225210c. Added explicit `[tool.setuptools.package-data]` entries for `meds_torchdata.preprocessing.tokenization_examples` / `...tensorization_examples` / `...configs` covering `**/*.yaml`. Verified with `uv build` that both `tokenization_examples/*.yaml` and `tensorization_examples/*.yaml` ship in the wheel. Also added `__init__.py` markers in both example directories to make the subpackages explicit rather than relying on setuptools-scm's implicit git-tracked-file inclusion.

(Worth noting: the YAMLs *did* already ship via setuptools-scm alone — confirmed via `unzip -l` on the built wheel before my fix — but that's fragile and Copilot was right that it shouldn't depend on implicit behavior.)

File: src/meds_torchdata/preprocessing/tensorization.py [hunk 2]
```diff
@@ -147,7 +150,11 @@ def convert_to_NRT(df: pl.LazyFrame) -> JointNestedRaggedTensorDict:
     return JointNestedRaggedTensorDict(tensors_dict, schema=schema)
 
 
-@Stage.register(is_metadata=False)
+@Stage.register(
+    is_metadata=False,
+    example_class=MTDStageExample,
+    examples_dir=Path(__file__).parent / "tensorization_examples",
+)
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/preprocessing/tensorization.py]: Same fix as on the tokenization thread — covered by the `[tool.setuptools.package-data]` block in `pyproject.toml` (225210c), plus the `__init__.py` marker in `tensorization_examples/`.

File: src/meds_torchdata/preprocessing/_stage_example.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        want_data_fp = example_dir / "out_data.yaml"
+        in_fp = example_dir / "in.yaml"
+        stage_cfg_fp = example_dir / "cfg.yaml"
+
+        return cls(
+            stage_name=stage_name,
+            scenario_name=scenario_name,
+            want_data=want_data_fp if want_data_fp.is_file() else None,
+            in_data=in_fp if in_fp.is_file() else None,
+            stage_cfg=safe_load(stage_cfg_fp.read_text()) if stage_cfg_fp.is_file() else {},
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/preprocessing/_stage_example.py]: Applied in 225210c. `cfg.yaml` is now loaded with `safe_load(...) or {}` and validated with an `isinstance(dict)` check that raises a `TypeError` with a clear message when the file isn't a mapping.

File: src/meds_torchdata/preprocessing/_stage_example.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            return
+
+        if is_resolved_dir:
+            raise NotImplementedError(
+                "MTDStageExample does not support is_resolved_dir=True — our shard paths "
+                "are declared relative to the cohort root (e.g. 'schemas/train/0.parquet'), "
+                "not to a single resolved `data/` or `metadata/` subdirectory."
+            )
+
+        spec = safe_load(self.want_data.read_text())
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/preprocessing/_stage_example.py]: Applied in 225210c. The `safe_load` result for `out_data.yaml` is now guarded by an `isinstance(spec, dict)` check that raises a descriptive `AssertionError` ("must contain a top-level mapping from shard-relative paths…") instead of the raw `AttributeError` on `spec.items()`.

File: src/meds_torchdata/preprocessing/tokenization.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/preprocessing/tokenization.py]: Why do these need to ship with the wheel? These are tests, not production or library code, right? Push back if I'm misunderstanding.

File: src/meds_torchdata/preprocessing/tensorization.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/preprocessing/tensorization.py]: same here as the other example -- do these need to ship with the wheel?

File: src/meds_torchdata/preprocessing/tokenization.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/preprocessing/tokenization.py]: Also, actually, a bigger point -- we should just structure our directories to be what is the expected default structure for MEDS transforms stages. Then this is picked up automatically. look at how MEDS transforms structures its stages. That will ensure MEDS-transforms' automatic docgen also works properly.

[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/preprocessing/tokenization.py]: You're right — they don't strictly need to ship. I dropped the explicit `[tool.setuptools.package-data]` config (commit d4bc564). setuptools-scm still auto-includes them from git-tracked files, which matches upstream's MEDS-Transforms behavior (upstream doesn't configure package-data either; their `stages/*/examples/*.yaml` ship via the same auto-inclusion). Net result: the wheel is unchanged, we just no longer have redundant packaging config to maintain.

If you want them *excluded* from the wheel entirely we can add explicit excludes — but that'd be diverging from upstream's convention for no practical gain (the files are a few KB total and the library doesn't rely on them at runtime). Happy to go either way; let me know.

File: src/meds_torchdata/preprocessing/tensorization.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/preprocessing/tensorization.py]: Same answer as on the tokenization thread — the explicit `[tool.setuptools.package-data]` is gone in d4bc564; setuptools-scm handles inclusion the same way upstream does.

File: src/meds_torchdata/preprocessing/tokenization.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/preprocessing/tokenization.py]: Done in d4bc564. Restructured both stages to mirror the upstream layout (e.g. `MEDS_transforms/stages/reshard_to_split/`):

```
preprocessing/
  tokenization/
    __init__.py          # `from .tokenization import main`
    tokenization.py      # the @Stage.register'd function
    examples/
      __init__.py
      in.yaml
      out_data.yaml
  tensorization/         # same shape
    ...
```

With this layout, `Stage.register` auto-infers `examples_dir` (the "calling file lives in a directory of the same name as the stage AND has an `examples` subdir" rule), so both decorators collapse to `@Stage.register(is_metadata=False, example_class=MTDStageExample)` — no explicit path argument. The `[project.entry-points."MEDS_transforms.stages"]` paths don't change because each subpackage's `__init__.py` re-exports `main`. Tests pass (3 stage tests + full suite 28 passed, 2 skipped). MEDS-Transforms' auto-docgen should pick us up via the same code path it uses for upstream stages.

File: tests/preprocessing/test_stages.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+our stages' singleton scenarios are already aligned to chain with each other.
+"""
+
+import pytest
+from MEDS_transforms.pytest_plugin import pipeline_tester
+
+
+def test_stage_example(stage_example):
+    stage_example.test()
+
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/preprocessing/test_stages.py:25]: Fair — the description was stale from the prototype commit. Rewrote the PR body to reflect the current state (old harness pruned, new scenarios added).

File: src/meds_torchdata/preprocessing/tokenization/examples/default/in.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    - 2020-01-01 00:00:00
+    - 2020-01-01 00:00:00
+    - 2020-01-11 00:00:00
+    - null
+    - null
+    - 2021-06-01 00:00:00
+    - 2021-06-01 00:00:00
+    - 2021-06-03 00:00:00
+  code: [5, 10, 11, 12, 6, 7, 20, 21, 22]
+  numeric_value: [1.0, 0.5, 0.7, 1.2, 2.0, 3.0, 5.5, 5.7, 6.2]
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/preprocessing/tokenization/examples/default/in.yaml]: Addressed in 3300a84. Added two new scenarios alongside the baseline `default/`:

- `tokenization/examples/missing_static/` — two-subject shard where subject 2 has no static rows, producing `null` in `static_code` / `static_numeric_value` columns. Exercises the same full-outer-join null-surfacing path the old `NORMALIZED_TRAIN_0_MISSING_STATIC` fixture covered.
- `tokenization/examples/multi_shard/` and `tensorization/examples/multi_shard/` — four shards across three splits (`train/0`, `train/1`, `tuning/0`, `held_out/0`, five subjects total). Exercises that the stages iterate correctly over arbitrarily-nested shard paths.

Both added purely as YAML under the upstream per-stage `examples/<scenario>/` layout — no custom test harness. Auto-parametrized across `test_stage_example[tokenization-missing_static]`, `test_stage_example[tokenization-multi_shard]`, `test_stage_example[tensorization-multi_shard]` via the `stage_example` fixture. All six stage test runs + the pipeline test pass.

File: tests/preprocessing/test_stages.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    """Same pipeline, run through joblib — replaces bespoke `test_preprocess_parallel.py`.
+
+    Gated on the `parallelized` marker so CI skips when `hydra-joblib-launcher` isn't
+    installed (the existing optional dep used for the old parallel test).
+    """
+    pipeline_tester(
+        pipeline_yaml=_PIPELINE_YAML,
+        stage_runner_yaml="parallelize:\n  n_workers: 2\n  launcher: joblib\n",
+        stage_scenario_sequence=["tokenization", "tensorization"],
+    )
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/preprocessing/test_stages.py:59]: Good call. Added `test_preprocess_stage_runner_fp_passthrough` to `test_preprocess.py` in 3300a84 — a ~20-line CLI test that invokes `MTD_preprocess stage_runner_fp=...` with a minimal joblib runner YAML and asserts the run produces NRT outputs. That covers the `if stage_runner_fp: cmd.extend(["--stage_runner_fp", str(stage_runner_fp)])` passthrough in `__main__.py` that `pipeline_tester` (which calls `MEDS_transform-pipeline` directly) bypasses. Gated on `@pytest.mark.parallelized` since it needs `hydra-joblib-launcher` — same gate as the old `test_preprocess_parallel.py`.

File: tests/preprocessing/test_stages.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
+"""Automated stage tests driven by `@Stage.register(examples_dir=...)`.
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/preprocessing/test_stages.py]: Reworded in 9792d08. The module docstring now says "Automated stage tests driven by registered `StageExample` scenarios." and adds a sentence clarifying that the `preprocessing/<stage>/<stage>.py` + adjacent `<stage>/examples/` layout is what lets `Stage.register` auto-infer `examples_dir` (no explicit argument). Thanks for catching the stale wording.

File: src/meds_torchdata/preprocessing/_stage_example.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    try:
+        assert_frame_equal(got, want, **kwargs)
+    except AssertionError as e:
+        raise AssertionError(
+            f"Parquet output {rel_path} differs from expected.\nGot:\n{got}\nWant:\n{want}"
+        ) from e
+
+
+def _check_nrt(rel_path: str, actual_fp: Path, want_tensors: dict) -> None:
+    got_nrt = JointNestedRaggedTensorDict(tensors_fp=actual_fp)
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/preprocessing/_stage_example.py]: Applied in 9792d08. `_check_nrt` now asserts `isinstance(want_tensors, dict)` up front and raises an `AssertionError` naming `rel_path` and the unexpected type before touching `JointNestedRaggedTensorDict(want_tensors)`. Same treatment on `_check_parquet` per the sibling thread.

File: src/meds_torchdata/preprocessing/_stage_example.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+
+def _check_parquet(rel_path: str, actual_fp: Path, want_cols: dict, df_check_kwargs: dict) -> None:
+    got = pl.read_parquet(actual_fp)
+    want = pl.DataFrame(want_cols)
+    # YAML doesn't natively express polars dtypes (u32 vs i64, f32 vs f64, etc.), so leave
+    # `check_dtypes` off by default and let callers override via `df_check_kwargs` when they
+    # want strict dtype validation. Value comparison + structural shape is the contract.
+    kwargs = {"check_column_order": False, "check_dtypes": False, **df_check_kwargs}
+    try:
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply src/meds_torchdata/preprocessing/_stage_example.py]: Applied in 9792d08. `_check_parquet` now asserts `isinstance(want_cols, dict)` before constructing `pl.DataFrame(want_cols)` and raises an `AssertionError` that names `rel_path` plus the unexpected type, so a malformed `out_data.yaml` surfaces an actionable message instead of a raw Polars construction error.

File: tests/preprocessing/test_preprocess.py [hunk 11]
```diff
@@ -96,6 +74,38 @@ def test_preprocess_path_with_spaces(simple_static_MEDS: Path):
...
+            f"MEDS_dataset_dir={simple_static_MEDS!s}",
+            f"output_dir={cohort_dir!s}",
+            f"stage_runner_fp={runner_fp!s}",
+        ]
+        out = subprocess.run(command, shell=False, check=False, capture_output=True, text=True)
+
+        assert out.returncode == 0, (
+            f"MTD_preprocess with stage_runner_fp failed (rc={out.returncode}).\n"
+            f"stdout:\n{out.stdout}\nstderr:\n{out.stderr}"
+        )
```
[the remark this answers is not in the document]
Comment (mmcdermott) [reply tests/preprocessing/test_preprocess.py:111]: Good call — the happy-path assertion would indeed pass if the wrapper silently dropped `stage_runner_fp` (the pipeline would just fall back to serial mode and still produce NRT outputs). Rather than fish log lines out of a hydra subprocess, I made the test prove passthrough by construction in 9792d08:

- Writes a runner YAML with `parallelize.launcher: __mtd_passthrough_sentinel_launcher__` — a launcher name no installed package provides.
- Invokes `MTD_preprocess` with `stage_runner_fp=<that-yaml>`.
- Asserts `returncode != 0` **and** that the sentinel launcher name appears in the captured stderr.

If passthrough works, the inner `MEDS_transform-pipeline` tries to resolve the sentinel launcher and fails with "Could not find 'hydra/launcher/__mtd_passthrough_sentinel_launcher__'" — the error surfaces both the failure and the sentinel string, so the test passes. If passthrough is broken (the wrapper drops `--stage_runner_fp`), the pipeline runs serially and succeeds → `returncode == 0` fails the test. No log-scraping needed.

Side benefit: the test no longer needs `hydra-joblib-launcher`, so I dropped the `@pytest.mark.parallelized` gate — it's now an always-on assertion of the wrapper's CLI contract. `test_stages.py::test_pipeline_parallel` retains the joblib-gated end-to-end coverage.

File: src/meds_torchdata/preprocessing/tokenization/examples/default/in.yaml
Comment (mmcdermott) [inline src/meds_torchdata/preprocessing/tokenization/examples/default/in.yaml:1]: Why do you need separate default and multi-shard examples? Can they not be the same example? What advantage does a single-shard example provide?

File: src/meds_torchdata/preprocessing/_stage_example.py
Comment (mmcdermott) [inline src/meds_torchdata/preprocessing/_stage_example.py:1]: I think you can simplify this code by using https://github.com/mmcdermott/yaml_to_disk explicitly. You can even register your own file type extensions through the supported syntax in this PR for nrts, then use yaml_to_disk to fully materialize the specified out data onto a temporary dir to do your comparisons, without needing any specific parse code for parquet files (already supported by yaml_to_disk) and after you add a nrt extension for nrt as well. Can you assess whether that would reduce the lines of code here and simplify?

File: src/meds_torchdata/preprocessing/tokenization/examples/default/in.yaml
Comment (mmcdermott) [reply src/meds_torchdata/preprocessing/tokenization/examples/default/in.yaml:1]: Good point — no real benefit from having both. Consolidated in b0746da: deleted the old single-shard `default/` for both stages and renamed `multi_shard/` → `default/`. The baseline scenario now exercises multi-shard iteration (four shards across three splits for tokenization, four shards for tensorization), and `missing_static/` remains as the edge-case. `test_stages.py`'s `stage_scenario_sequence` still points at `tokenization/default` / `tensorization/default`, so no test-side changes were needed. Net: -78 lines of duplicated YAML.

File: src/meds_torchdata/preprocessing/_stage_example.py
Comment (mmcdermott) [reply src/meds_torchdata/preprocessing/_stage_example.py:1]: I prototyped this on a throwaway branch to measure. Findings — it's LOC-neutral-ish, not a clear reduction:

**What the refactor would do**
- Add `_nrt_file.py` with a `NRTFile(FileType)` subclass (~16 code lines + docstring/doctest).
- Register it via `[project.entry-points."yaml_to_disk.file_types"] nrt = ...` in `pyproject.toml`.
- Add `yaml_to_disk[parquet]` to `[project].dependencies` (noting `pyarrow` is already transitive via polars/meds, and yaml_to_disk itself is actually already transitive via `meds-transforms`).
- Rewrite `MTDStageExample.check_outputs` to `with yaml_disk(self.want_data) as expected_root:` and then `.rglob` the materialized tree, comparing file-by-file.
- Replace `_check_parquet` / `_check_nrt` with a single `_compare(expected_fp, actual_fp, rel, df_check_kwargs)` that reads both sides from disk and does format-aware comparison (`assert_frame_equal` for parquet, per-tensor `np.array_equal(equal_nan=True)` for NRT).

**Code-LOC (non-blank, non-comment, docstrings stripped via `ast`)**
- Current `_stage_example.py`: **73 lines**
- Refactored `_stage_example.py` (67) + `_nrt_file.py` (16): **83 lines**

So strictly by lines of code it's a small *increase*, not a reduction. The real wins are structural:
1. **Symmetric comparison** — both sides come off disk via their real on-disk reps, so there's no asymmetry between "spec-in-memory" and "actual-on-disk".
2. **Format validation by construction** — yaml_to_disk's `FileType.validate` handles the isinstance-checks that Copilot flagged earlier.
3. **`NRTFile` is reusable** — any downstream project using yaml_to_disk + NRT would get `.nrt:` yaml support for free.

**What I did**
- Kept the consolidation (separate comment, b0746da).
- Reverted the yaml_to_disk refactor from my working copy since the LOC story isn't compelling — want your call before committing.

Two ways forward if you still want to pursue:
- **(a) Keep it local.** Do the refactor as described and accept the small LOC increase for the structural wins. `NRTFile` lives in `meds_torchdata.preprocessing._nrt_file`.
- **(b) Upstream `NRTFile` into `nested_ragged_tensors`** (or yaml_to_disk itself) — the file-type registration logically belongs with NRT's home package, and then meds-torch-data is a pure consumer. More coordination but cleaner ownership.

Happy to execute either — let me know which (or skip both).

Comment (mmcdermott) [reply src/meds_torchdata/preprocessing/_stage_example.py:1]: Good call on both counts — pushed as 4792530 (yaml_to_disk refactor) + a396817 (set-based presence check).

**Set-based presence check (a396817).** Collapsed the iterate-and-raise pattern into:

```python
expected = {_canonical(fp.relative_to(expected_root)): fp for fp in expected_root.rglob("*") if ...}
actual   = {fp.relative_to(output_dir): fp for top in {rel.parts[0] for rel in expected} for fp in (output_dir / top).rglob("*") if ...}
assert expected.keys() == actual.keys(), f"Missing: ...\nUnexpected: ..."
for rel in sorted(expected):
    _compare(expected[rel], actual[rel], rel, self.df_check_kwargs)
```

`_ALLOWED_SUFFIXES = frozenset({".parquet", ".nrt"})` at module scope. One quirk worth flagging: I restrict the `actual` scan to the top-level dirs present in `expected` (e.g., `data/`) rather than `output_dir.rglob` everywhere. Otherwise in standalone mode `output_dir` is the cohort root and we'd sweep up `metadata/codes.parquet` / `metadata/subject_splits.parquet` as "unexpected" — those are upstream MEDS artifacts our stages don't produce. Happy to loosen this if you'd prefer a stricter contract.

**`__eq__` + NaN.** Filed the upstream issue as [nested_ragged_tensors#63](https://github.com/mmcdermott/nested_ragged_tensors/issues/63) — repro + proposed fix is just adding `equal_nan=True` to the `np.array_equal` call inside `__eq__`. Once that lands we can replace the NRT branch of `_compare` with `assert want == got`, which collapses ~14 lines to ~2 and tips the overall refactor net-negative on LOC.

**Moving `NRTFile` upstream.** Agreed it belongs in nested_ragged_tensors (or yaml_to_disk). Left it local for now so this PR stays self-contained; happy to do the upstream move as a follow-up once this lands.

**Current LOC math (code-only, docstrings stripped via ast):**
- Before: 73
- After: 64 (`_stage_example.py`) + 16 (`_nrt_file.py`) = 80
- Net: +7 lines, pending the `__eq__` upstream fix that will take it to -5.

File: tests/preprocessing/__init__.py
Comment (mmcdermott) [inline tests/preprocessing/__init__.py:1]: you no longer need this file. Just declare the string constant where it is needed. I think it is only used once, right?

File: tests/preprocessing/test_stages.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+
+def test_stage_example(stage_example):
+    stage_example.test()
+
+
+_PIPELINE_YAML = "input_dir: {input_dir}\noutput_dir: {output_dir}\nstages: [tokenization, tensorization]\n"
+
+
+def test_pipeline_serial():
```
Comment (mmcdermott) [inline tests/preprocessing/test_stages.py:39]: IT's worth noting in a comment that these pipeline stages _don't_ test the built in MTD CLI entry point. Instead, they test the internals of that but strictly through meds-transforms entry points. This is an awkward kind of testing indirection that we should eventually try to solve, though I'm not sure if it is worth trying to change here. Though do check and explore if there is anything in the pipeline_tester function that we could do to somehow make this better, just in case I'm wrong.

File: conftest.py [hunk 13]
```diff
@@ -1,4 +1,10 @@
-"""Test set-up and fixtures code."""
+"""Test set-up and fixtures code.
+
+`MEDS_transforms.pytest_plugin` is auto-registered as a pytest entry-point plugin
```
Comment (mmcdermott) [inline conftest.py]: you don't need this comment.

File: src/meds_torchdata/preprocessing/_stage_example.py [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+            # type inference diverges from polars' for nested lists, so leave `check_dtypes`
+            # off by default and let callers override via `df_check_kwargs`.
+            kwargs = {"check_column_order": False, "check_dtypes": False, **df_check_kwargs}
+            got = pl.read_parquet(actual_fp)
+            want = pl.read_parquet(expected_fp)
+            try:
+                assert_frame_equal(got, want, **kwargs)
+            except AssertionError as e:
+                raise AssertionError(f"Parquet {rel} differs.\nGot:\n{got}\nWant:\n{want}") from e
+        case ".nrt":
```
Comment (mmcdermott) [inline src/meds_torchdata/preprocessing/_stage_example.py:121]: add a comment pointing to the issue you added in NRT about why we don't just use equality.

Comment (mmcdermott) [reply src/meds_torchdata/preprocessing/_stage_example.py:121]: Done in 727e0b2. Added a comment directly above the NRT compare branch linking [nested_ragged_tensors#63](https://github.com/mmcdermott/nested_ragged_tensors/issues/63), so future readers hitting "why not just `assert want == got`?" can trace straight back to the `equal_nan=True` gap.

File: conftest.py (hunk 13 above)
Comment (mmcdermott) [reply conftest.py]: Trimmed in 727e0b2 — `conftest.py`'s docstring is now just `"Test set-up and fixtures code."`. The plugin-discovery explanation was redundant with `test_stages.py`'s own docstring.

File: tests/preprocessing/test_stages.py (hunk 12 above)
Comment (mmcdermott) [reply tests/preprocessing/test_stages.py:39]: Added the note in 727e0b2. And I did dig into `pipeline_tester` to check whether it could route through `MTD_preprocess` instead — it can't, at least not without an upstream change:

```python
# MEDS_transforms/pytest_plugin.py inside pipeline_tester
command = ["MEDS_transform-pipeline", str(pipeline_config_fp)]
...
out = run_fn(command, shell=False, capture_output=True)
```

The `run_fn` parameter lets you swap the subprocess runner (useful for unit-testing `pipeline_tester` itself, as upstream does in its doctest), but the `command` list is hard-coded. So there's no hook to say "drive this through my `MTD_preprocess` wrapper instead". The test_stages.py comment now explicitly flags this as something worth revisiting if upstream adds a command-override knob.

Happy to file an upstream issue on MEDS_transforms proposing a `command_prefix=` or similar — it'd be a small addition (swap the hardcoded list for a caller-supplied one defaulting to the current value) and would let downstream wrappers like `MTD_preprocess` fold back into the same pipeline tests. Say the word and I'll write it up.

File: tests/preprocessing/__init__.py
Comment (mmcdermott) [reply tests/preprocessing/__init__.py:1]: Deleted in 727e0b2. `PREPROCESS_SCRIPT = "MTD_preprocess"` is now declared inline at the top of `test_preprocess.py` (the only file that referenced it), and `tests/preprocessing/__init__.py` is gone.
